src/pool/srv_pool.c

```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
+		if (rc1 != 0)
+			return rc1;
+	}
+
+	rc1 = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc1 != 0)
+		D_GOTO(out_svc, rc1);
+
+	ABT_rwlock_wrlock(svc->ps_lock);
+	if (rc == 0)
+		val = DAOS_UPGRADE_STATUS_COMPLETED;
+	else
+		val = DAOS_UPGRADE_STATUS_FAILED;
+
+	d_iov_set(&value, &val, sizeof(val));
+	rc1 = rdb_tx_update(&tx, &svc->ps_root, &ds_pool_prop_upgrade_status,
+			   &value);
+	if (rc1)
+		D_GOTO(out_tx, rc1);
+
+	/* only bump global version if upgrade succeed. */
+	if (rc == 0) {
+		val = DS_POOL_GLOBAL_VERSION;
+		rc1 = rdb_tx_update(&tx, &svc->ps_root,
+				    &ds_pool_prop_global_version, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade global version "
+				"of pool, %d.\n", DP_UUID(pool_uuid), rc1);
+			D_GOTO(out_tx, rc1);
+		}
+	}
+
+	rc1 = pool_prop_read(&tx, svc, DAOS_PO_QUERY_PROP_ALL, &prop);
+	if (rc1)
+		D_GOTO(out_tx, rc1);
+	rc1 = ds_pool_iv_prop_update(svc->ps_pool, prop);
+	daos_prop_free(prop);
+	if (rc1)
+		D_GOTO(out_tx, rc1);
+
+	rc1 = rdb_tx_commit(&tx);
+out_tx:
+	ABT_rwlock_unlock(svc->ps_lock);
+	rdb_tx_end(&tx);
+out_svc:
+	if (need_put_leader)
+		pool_svc_put_leader(svc);
+
+	return rc1;
+}
+
+static int
+ds_pool_upgrade_if_needed(uuid_t pool_uuid, struct rsvc_hint *po_hint,
+			  struct pool_svc *svc, crt_rpc_t *rpc)
+{
+	struct rdb_tx			tx;
+	d_iov_t				value;
+	uint32_t			upgrade_status;
+	uint32_t			upgrade_global_ver;
+	int				rc, rc1;
+	bool				upgraded = false;
+	bool				need_put_leader = false;
+
+	if (!svc) {
+		rc = pool_svc_lookup_leader(pool_uuid, &svc, po_hint);
+		if (rc != 0)
+			return rc;
+		need_put_leader = true;
+	}
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc != 0)
+		D_GOTO(out_svc, rc);
+
+	/**
+	 * Four kinds of pool upgrading states:
+	 *
+	 * 1. pool upgrade not started:
+	 * upgrade_state: not started
+	 * upgrade_global_version: v1
+	 * global_version: v1
+	 *
+	 * 2. pool upgrade in progress:
+	 * upgrade_state: in progress
+	 * upgrade_global_version: v2
+	 * global_version: v1
+	 *
+	 * 3. pool upgrade completed:
+	 * upgrade_state: completed
+	 * upgrade_global_version: v2
+	 * global_version: v2
+	 *
+	 * 4. pool upgrade failed:
+	 * upgrade_state: failed
+	 * upgrade_global_version: v2
+	 * global_version: v1
+	 */
+	ABT_rwlock_wrlock(svc->ps_lock);
+	d_iov_set(&value, &upgrade_global_ver, sizeof(upgrade_global_ver));
+	rc = rdb_tx_lookup(&tx, &svc->ps_root, &ds_pool_prop_upgrade_global_version,
+			   &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_tx, rc);
+	} else if (rc == -DER_NONEXIST) {
+		if (!need_put_leader)
+			D_GOTO(out_tx, rc = 0);
+		D_GOTO(out_upgrade, rc);
+	} else {
+		d_iov_set(&value, &upgrade_status, sizeof(upgrade_status));
+		rc = rdb_tx_lookup(&tx, &svc->ps_root, &ds_pool_prop_upgrade_status,
+				   &value);
+		if (rc)
+			D_GOTO(out_tx, rc);
+
+		if (upgrade_global_ver > DS_POOL_GLOBAL_VERSION) {
+			D_ERROR("downgrading pool is unsupported\n");
```

Comment:
[Nit] All these error messages should be in the format 'DF_UUID": <msg>\n"' so that we know which pool has the problem and what the problem is. For instance, this one should be

    DF_UUID": downgrading pool not supported: %u -> %u\n", DP_UUID(svc->ps_uuid)

Reply:
Sure